src/main/java/redis/clients/jedis/JedisSocketFactory.java

```diff
@@ -24,6 +24,8 @@
    */
   Socket createSocket() throws IOException, JedisConnectionException;
 
+  JedisSocketFactory copyWith(HostAndPort hostAndPort);
```

Comment:
I'm not that happy with this one. What could be other options?

Frankly, I was working with this very PR while you pushed the resolution of conflicts.
My very last thought process was to use a completely new JedisSocketFactory containing `Socket createSocket(HostAndPort)`.

Reply:
> I'm not that happy with this one. What could be other options?

I agree, it's an annoying one and I couldn't (yet?) find a perfect solution here

> Frankly, I was working with this very PR while you pushed the resolution of conflicts.

Sorry for that, it just had too many changes after rebasing on top of your earlier config changes, so I had to abandon the branch and force a new one.

> My very last thought process was to use a completely new JedisSocketFactory containing `Socket createSocket(HostAndPort)`.

That would kind of break the concept as for UDS there is no real "host and port", though maybe that's okay since UDS is a very rare usecase but in that case we would need to also store a reference to hostandport in addition to the socket factory which is not the best thing to do IMO.

Reply:
One alternative is to deprecate the setHostAndPort (currently only used in the sentinel pool I think and cleanup that completely)

Reply:
> Sorry for that, ...

There's nothing to be sorry about :)

> ... deprecate the setHostAndPort ...

We have to keep sentinel under consideration as well, where master (HostAndPort) can be changed. If not with this pattern, may be with a different pattern.

Reply:
> We have to keep sentinel under consideration as well, where master (HostAndPort) can be changed. If not with this pattern, may be with a different pattern.

I didn't mean kill the sentinel at all (though it looks like Redis 7 might retire it? https://github.com/redis/redis/projects/4), I meant replace that with something else that's more manageable (though maybe that's just moving the problem around?).

I will have a think about it, but I am also open to suggestions here!

Reply:
> it looks like Redis 7 might retire it?

We are yet to retire sharding, sentinel would be a long shot :(

Reply:
Just another idea.
```suggestion
  default JedisSocketFactory updateHostAndPort(HostAndPort hostAndPort) {
    return this;
  }
```

Reply:
I don't object to the name, if we are okay with the JedisSocketFactory knowing explicitly about the hostandport then I think it's fine, I don't think we should provide a default here because that can lead to nasty bugs though, I would prefer it breaks compling if someone is already using the socket factory feature (which is a very niche feature and I don't think anyone is actually using it). I will add another commit with these changes and we can have a look.